text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+---
+feature name: cli-framework-compatibility-strategy
+start date: February 2nd, 2020
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/111
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/110
+---
+
+<!-- replace the blockquoted sections with your content -->
+# Summary
+
+There are currently two promises we try to keep with regards to framework and cli compatibility:
+
+- CLI versions should always support older framework versions. That is, upgrading the CLI should never require an upgrade of the framework.
+- Framework versions might require an upgrade of the CLI. In which case, we provide instructions as to which CLI version should be used.
+
+This RFC details the necessary strategy and mechanism for ensuring that we keep those promises when releasing new versions.
+
+# Motivation
+
+After implementing this RFC, we can expect the following outcomes:
+
+- An incompatible change in the CLI won't make it to the release.
+- An incompatible change in the framework, will always be accompanied with a protocol version bump.
+- An incompatible change in the framework, accompanied with a protocol version bump, will always produce the correct error message.
+
+There are 2 main motivations why we should put some substantial effort into this:
+
+## (1) Haven't we suffered enough?
+
+Recently, we experienced several incidents that are directly related to this domain:
+
+- [aws/aws-cdk: A newer version of the CDK CLI (>= 1.21.0) is necessary to interact with this app](https://github.com/aws/aws-cdk/issues/5986)
```

Comment:
I agree that we have had issues that produce error messages related to versioning. I'm not sure that we've root-caused these enough to be able to say for certain that they're all due to the same thing or their causes will be addressed by what's written in this RFC.  I'd like to see some motivation around that rather than listing issues.

Afaict:

* 1st one: caused by accidental caret dependency
* 2nd one: not root caused, simply closed as "fixed itself"
* 3rd one: caused by accidental caret dependency

In fairness, I haven't read the rest of the RFC yet.

Reply:
> 1st one: caused by accidental caret dependency

Not exactly. It's true that the error itself was thrown because of a caret dependency, but actually the bigger problem is that the error **message** is wrong, and instructs users to install an already installed CLI version. This problem would have been detected by the tests proposed in this RFC.

> 2nd one: not root caused, simply closed as "fixed itself"

Your'e right, this can be removed I guess.

> 3rd one: caused by accidental caret dependency

Well, the caret dependency only caused a side-effect, which was that users were unable to escape the problem even if they downgraded their CLI. 

*From the COE*:

```text
CDK CLI can only be used with apps created by CDK >= 1.10.0
```

*"The message was not clear at all, and many users did not know how to proceed from this point. Even those that did, and were NPM-savvy enough to know how to install an old CLI version could not escape from this situation. Because of the way the CLI specified a dependency on the module that did the actual version check:"*

```json
"dependencies: {
  "@aws-cdk/cx-api": "^1.9.0"
}
```

The actual problem was that the new CLI version simply didn't support older framework versions. In fact, this issue is what eventually made us implement the `upgradeAssemblyManifest` mechanism is this PR: https://github.com/aws/aws-cdk/pull/4307/files

The mechanism proposed in this RFC would have flushed this out because it doesn't allow for CLI versions to be incompatible with older framework versions.